File: ChangeTheme_OC/ChangeTheme_OC/UIStyle/ArtUIStyle.m
//
//  ArtUIStyle.m
//  ChangeTheme_OC
//
//  Created by weijingyun on 2017/8/21.
//  Copyright © 2017年 weijingyun. All rights reserved.
//

#import "ArtUIStyle.h"
#import "ArtUIStyleManager.h"
#import "UIColor+HexColor.h"


#pragma mark - ArtUIStyle

@interface ArtUIStyle ()

@property (nonatomic, assign) CGFloat top;
@property (nonatomic, assign) CGFloat left;
@property (nonatomic, assign) CGFloat bottom;
@property (nonatomic, assign) CGFloat right;
@property (nonatomic, assign) CGFloat width;
@property (nonatomic, assign) CGFloat height;
@property (nonatomic, assign) CGFloat centerXOffset;
@property (nonatomic, assign) CGFloat centerYOffset;


@property (nonatomic, strong) UIColor *styleColor;
@property (nonatomic, strong) UIFont *styleFont;
@property (nonatomic, strong) UIImage *styleImage;

@property (nonatomic, strong) NSDictionary* style;

@end

@implementation ArtUIStyle

- (id)initWithStyle:(NSDictionary *)aStyle
{
    if (self = [super init]) {
        _style = aStyle;
        [self setValuesForKeysWithDictionary:aStyle];
    }
    return self;
}

- (instancetype)initWithStyle:(NSDictionary *)aStyle imageString:(NSString *)aImageString{
    if (self = [super init]) {
        _style = aStyle;
        self.styleImage = [self imageString:aImageString];
    }
    return self;
}

- (void)setValue:(id)value forUndefinedKey:(NSString *)key {

}


- (UIFont *)styleFont
{
    if (!_styleFont) {
        NSNumber *num = [self.style objectForKey:kArtUIStyleFontKey];
        NSAssert(num != nil, @"配置的字体大小不存在请检查");
        _styleFont = [UIFont systemFontOfSize:[num doubleValue]];
    }
    return _styleFont;
}

- (UIColor *)styleColor
{
    if (!_styleColor) {
        NSString *colorStr = [self.style objectForKey:kArtUIStyleColorKey];
        colorStr = [colorStr stringByReplacingOccurrencesOfString:@" " withString:@""];
        NSAssert(colorStr.length > 0, @"配置的颜色值不存在请检查");
        NSArray<NSString *> *colorArray = [colorStr componentsSeparatedByString:@","];
        NSString *hexStr = colorArray.firstObject;
        CGFloat alpha = 1.0;
        if (colorArray.count == 2) {
            alpha = [colorArray.lastObject doubleValue];
        }
        _styleColor = [UIColor art_colorWithHexString:hexStr alpha:alpha];
        
    }
    return [_styleColor copy];
}

- (UIImage *)styleImage {
    return [_styleImage copy];
}

- (UIImage *)imageString:(NSString *)aImageString {
    
    ArtUIStyleManager *manager = [ArtUIStyleManager shared];
    NSString *toPath = self.style[@"toPath"];
    UIImage *image = nil;
    switch (manager.styleType) {
        case EArtUIStyleTypeDefault:
        {
             image = [UIImage imageNamed:aImageString];
        }
            break;
            
        case EArtUIStyleTypeBundle:
        {
            image =
            [self findImageForImageString:aImageString block:^NSString *(NSString *path) {
                return [[NSBundle bundleWithPath:manager.stylePath] pathForResource:path ofType:@"png" inDirectory:toPath];
            }];
            
        }
            break;
            
        case EArtUIStyleTypeStylePath:
        {
            image =
            [self findImageForImageString:aImageString block:^NSString *(NSString *path) {
                NSString *filePath = [[manager.stylePath stringByAppendingPathComponent:toPath] stringByAppendingPathComponent:[NSString stringWithFormat:@"%@.png",path]];
                return filePath;
            }];
        }
            break;
            
        default:
            break;
    }
    return image;
}

- (UIImage *)findImageForImageString:(NSString *)aImageString block:(NSString *(^)(NSString * path))aBlock {
    NSMutableArray *arrayM = [NSMutableArray new];
    for (int i = 1; i <= 3; i ++) {
        [arrayM addObject:@(i)];
    }
    NSInteger scale = (NSInteger)[UIScreen mainScreen].scale;
    [arrayM removeObject:@(scale)];
    [arrayM insertObject:@(scale) atIndex:0];
    
    __block UIImage *image = nil;
    [arrayM enumerateObjectsUsingBlock:^(NSNumber *obj, NSUInteger idx, BOOL * _Nonnull stop) {
        NSInteger scale = obj.integerValue;
        NSString *imageStr = aImageString;
        if (scale != 1) {
            imageStr = [NSString stringWithFormat:@"%@@%tux",aImageString,scale];
        }
        NSString *imagePath = aBlock(imageStr);
        image = [UIImage imageWithContentsOfFile:imagePath];
        if (image) {
            *stop = YES;
        }
    }];
    if (image == nil) {
        NSAssert(NO, @"未在资源中找到可用的图");
        image = [UIImage imageNamed:aImageString];
    }
    return image;
}

@end


@implementation ArtUIStyle (ArtUIStyleApp)

+ (instancetype)artModule:(NSString *)aModule styleForKey:(NSString *)aStyleKey {
    return [[ArtUIStyleManager shared] artStyleModule:aModule styleKey:aStyleKey];
}

+ (void)artModule:(NSString *)aModule styleForKey:(NSString *)aStyleKey strongSelf:(id)strongSelf block:(void(^)(ArtUIStyle *style, id weakSelf))aBlock {
    if (strongSelf) {
        [[ArtUIStyleManager shared] saveStrongSelf:strongSelf block:^(id weakSelf) {
            ArtUIStyle *style = [self artModule:aModule styleForKey:aStyleKey];
            aBlock(style,weakSelf);
        }];
    }
}

@end

@implementation UIColor (ArtUIStyleApp)

+ (UIColor *)artModule:(NSString *)aModule colorForKey:(NSString *)aColorKey {
    return [ArtUIStyle artModule:aModule styleForKey:aColorKey].styleColor;
}

+ (void)artModule:(NSString *)aModule colorForKey:(NSString *)aColorKey strongSelf:(id)strongSelf block:(void(^)(UIColor *color, id weakSelf))aBlock {
    if (strongSelf) {
        [[ArtUIStyleManager shared] saveStrongSelf:strongSelf block:^(id weakSelf) {
            UIColor *color = [self artModule:aModule colorForKey:aColorKey];
            aBlock(color,weakSelf);
        }];
    }
}

@end

@implementation UIFont (ArtUIStyleApp)

+ (UIFont *)artModule:(NSString *)aModule fontForKey:(NSString *)aFontKey {
    return [ArtUIStyle artModule:aModule styleForKey:aFontKey].styleFont;
}

+ (void)artModule:(NSString *)aModule fontForKey:(NSString *)aFontKey strongSelf:(id)strongSelf block:(void(^)(UIFont *font, id weakSelf))aBlock {

    if (strongSelf) {
        [[ArtUIStyleManager shared] saveStrongSelf:strongSelf block:^(id weakSelf) {
            UIFont *font = [self artModule:aModule fontForKey:aFontKey];
            aBlock(font,weakSelf);
        }];
    }
}

@end


@implementation UIImage (ArtUIStyleApp)

+ (UIImage *)artModule:(NSString *)aModule imageString:(NSString *)aImageString {
    return [[ArtUIStyleManager shared] artImageModule:aModule imageString:aImageString].styleImage;
}

+ (void)artModule:(NSString *)aModule imageString:(NSString *)aImageString strongSelf:(id)strongSelf block:(void(^)(UIImage *image, id weakSelf))aBlock {
    
    if (strongSelf) {
        [[ArtUIStyleManager shared] saveStrongSelf:strongSelf block:^(id weakSelf) {
            UIImage *image = [self artModule:aModule imageString:aImageString];
            aBlock(image,weakSelf);
        }];
    }
}

@end

File: ChangeTheme_Swift/ChangeTheme_Swift/UIStyle/ArtUIStyle.swift
//
//  ArtUIStyle.swift
//  ChangeTheme_Swift
//
//  Created by weijingyun on 2017/9/5.
//  Copyright © 2017年 weijingyun. All rights reserved.
//

import UIKit

fileprivate let kArtUIStyleFontKey = "font"
fileprivate let kArtUIStyleColorKey = "color"

class ArtUIStyle: NSObject {
    
    fileprivate var saveImage : UIImage?
    fileprivate var saveColor : UIColor?
    fileprivate var saveFont  : UIFont?
    let top             = 0.0
    let left            = 0.0
    let bottom          = 0.0
    let right           = 0.0
    let height          = 0.0
    let centerXOffset   = 0.0
    let centerYOffset   = 0.0
    
    init(styleDic : [String : AnyObject]) {
        super.init()
        setValuesForKeys(styleDic)
        do {
            var colorStr = try styleDic[kArtUIStyleColorKey].unwrap(tip: kArtUIStyleColorKey + "不存在") as! String
            colorStr = colorStr.replacingOccurrences(of: " ", with: "")
            let colorArray = colorStr.components(separatedBy: ",")
            var alpha = 1.0
            if colorArray.count == 2 {
                alpha = Double(colorArray.last!)!
            }
            saveColor = UIColor.art_color(hexString: colorArray.first!, alpha: CGFloat(alpha))
        } catch {
//            print(error)
        }
        
        do {
            let font = try styleDic[kArtUIStyleFontKey].unwrap(tip: kArtUIStyleFontKey + "不存在") as! NSNumber
            saveFont = UIFont.systemFont(ofSize: CGFloat(font.doubleValue))
        } catch {
//            print(error)
        }
    }
    
    init(imageDic : [String : AnyObject], imageString : String) {
        super.init()
        let manager = ArtUIStyleManager.share
        let toPath = imageDic["toPath"] as! String
        do {
            
            switch manager.styleType {
            case .Default:
                saveImage = UIImage.init(named: imageString)
                break
                
            case .Bundle:
                let bundle = try Bundle.init(path: try manager.stylePath.unwrap()).unwrap()
                saveImage = findImage(imageString: imageString, block: { (path) -> (String?) in
                   return bundle.path(forResource: path, ofType: "png", inDirectory: toPath)
                })
                break
                
            case .Path:
                let stylePath = try manager.stylePath.unwrap()
                saveImage = findImage(imageString: imageString, block: { (path) -> (String?) in
                    return stylePath + "/" + toPath + "/" + path + ".png"
                })
                break
                
            }
            
        } catch {
            saveImage = UIImage.init(named: imageString)
        }
    }
    
    func findImage(imageString : String, block : (_: String) -> (String?)) -> UIImage {
        var arrayM = [String]()
        let scale = NSInteger(UIScreen.main.scale)
        arrayM.append(String.init(stringInterpolationSegment:scale))
        for index in 1...3 {
            if scale != index {
                arrayM.append(String.init(stringInterpolationSegment: index))
            }
        }
        
        var image : UIImage?
        for scaleStr in arrayM {
            var imageStr = imageString
            if NSInteger(scaleStr) != 1 {
                imageStr = imageStr + "@" + scaleStr + "x"
            }
            let imagePath = block(imageStr)
            if imagePath != nil {
                image = UIImage.init(contentsOfFile: imagePath!)
            }
            if image != nil {
                break;
            }
        }
   
        if image == nil {
            assert(false,"未在资源中找到可用的图")
            image = UIImage.init(named: imageString)
        }
        return image!
    }
    
    override func setValue(_ value: Any?, forUndefinedKey key: String) {
        
    }
    
    func color() -> UIColor {
        assert(saveColor != nil, "请检查")
        return saveColor!
    }
    
    func image() -> UIImage {
        assert(saveImage != nil, "请检查")
        return saveImage!
    }
    
    func font() -> UIFont {
        assert(saveFont != nil, "请检查")
        return saveFont!
    }
}

extension ArtUIStyle {
    class func artStyle(module : String, styleKey : String) -> ArtUIStyle {
        return ArtUIStyleManager.artStyle(module: module, styleKey:styleKey)
    }
}

extension UIColor {
    class func artStyle(module : String, styleKey : String) -> UIColor {
        return ArtUIStyleManager.artStyle(module: module, styleKey:styleKey).color()
    }
}

extension UIFont {
    class func artStyle(module : String, styleKey : String) -> UIFont {
        return ArtUIStyleManager.artStyle(module: module, styleKey:styleKey).font()
    }
}

extension UIImage {
    class func artImage(module : String, imageString : String) -> UIImage {
        return ArtUIStyleManager.artImage(module: module, imageString: imageString).image()
    }
}
